qiskit/ignis/measurement/discriminator/iq_discrimination.py

```diff
@@ -0,0 +1,264 @@
+            expected_state: the expected state of the discriminator.
+        """
+        self._expected_state.append(expected_state)
+        self._backend_result_list.append(result)
+
+        if refit:
+            self.fit_data()
+
+    def _scale_data(self, xdata: List[List[float]] = None):
+        """
+        Scales the features of xdata to have mean zero and unit variance.
+        Args:
+            xdata:
+        Returns:
+            data scaled to have zero mean and unit variance.
+        """
+        if not self._standardize:
+            return xdata
+
+        if not self._scaler:
+            self._scaler = StandardScaler(with_std=True)
+            self._scaler.fit(self._xdata)
+
+        if not xdata:
+            self._xdata = self._scaler.transform(self._xdata)
+            return
+
+        return self._scaler.transform(xdata)
+
+    def _calc_data(self):
+        """
+        Extract the measured IQ points (i.e. features) from the list of
+        backend results and combine them into a list of lists.
+        Returns:
+            x a list of lists. Each sublist corresponds to the IQ points
+            for all the qubits. Each sublist therefore corresponds to an
+            expected result stored in self._expected_state.
+        """
+        self._xdata, self._ydata = [], []
+        for schedule in self._circuit_names:
+            for result in self._backend_result_list:
+                try:
+                    iq_data = result.get_memory(schedule)[:, self._qubits]
+                    for shot_idx in range(iq_data.shape[0]):
+                        shot = np.array(iq_data[shot_idx])
+
+                        iq_shot = zip(list(np.real(shot)), list(np.imag(shot)))
+                        self._xdata.append([val for pair in iq_shot
+                                            for val in pair])
+
+                        if isinstance(schedule, Schedule):
+                            self._ydata.append(
+                                self._expected_state[schedule.name])
+                        else:
+                            self._ydata.append(self._expected_state[schedule])
+
+                except (QiskitError, KeyError):
+                    pass
+
+        self._scale_data()
+
+    def fit_data(self, qid=-1, p0=None, bounds=None, series=None):
+        """
+        Fit the Linear Discriminant.
+        Args:
+            qid: not needed
+            p0: not needed
+            bounds: not needed
+            series: not needed
+        """
+        self._fit_fun.fit(self._xdata, self._ydata)
+
+    def discriminate(self, x_data: List[List[float]]) -> List[str]:
+        """
+        Use the trained discriminator to discriminate the given data.
+        :param x_data: data properly formatted for the discriminator.
+        :return: the properly classified data.
+        """
+        return self._fit_fun.predict(x_data)
+
+    def extract_xdata(self, result: ExperimentResult):
```

Comment:
How does this handle the case of `meas_level=1`, `meas_return='single'`?

Reply:
Good catch. It handles `meas_return='single'` no problem but will fail when `meas_return='avg'`. I updated the code so that discriminator will work with `single` and `avg`.